src/category_theory/quotient.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2020 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+
+import category_theory.functor
+import category_theory.eq_to_hom
+import group_theory.free_group
+
+/-!
+# Quotient category
+
+Constructs the quotient of a category by an arbitrary family of relations on its hom-sets,
+by introducing a type synonym for the objects, and identifying homs as necessary.
+
+This is analogous to 'the quotient of a group by the normal closure of a subset', rather
+than 'the quotient of a group by a normal subgroup'.
+-/
+
+universes v v₁ u u₁
+
+namespace category_theory
+
+variables {C : Type u} [category.{v} C]
+          (r : Π (a b : C), (a ⟶ b) → (a ⟶ b) → Prop)
+include r
+
+/-- A type synonom for `C`, thought of as the objects of the quotient category. -/
+structure quotient := (to_C : C)
+
+instance [inhabited C] : inhabited (quotient r) := ⟨⟨r, default C⟩⟩
+
+namespace quotient
+
+/-- Generates the closure of a family of relations w.r.t. composition from left and right. -/
+inductive comp_closure {s t : C} : (s ⟶ t) → (s ⟶ t) → Prop
+| intro {a b} (f : s ⟶ a) (m₁ m₂ : a ⟶ b) (g : b ⟶ t) (h : r _ _ m₁ m₂) :
+  comp_closure (f ≫ m₁ ≫ g) (f ≫ m₂ ≫ g)
+
+lemma comp_left {a b c : C} (f : a ⟶ b) : Π (g₁ g₂ : b ⟶ c) (h : comp_closure r g₁ g₂),
+  comp_closure r (f ≫ g₁) (f ≫ g₂)
+| _ _ ⟨x, m₁, m₂, y, h⟩ := by simpa using comp_closure.intro (f ≫ x) m₁ m₂ y h
+
+lemma comp_right {a b c : C} (g : b ⟶ c) : Π (f₁ f₂ : a ⟶ b) (h : comp_closure r f₁ f₂),
+  comp_closure r (f₁ ≫ g) (f₂ ≫ g)
+| _ _ ⟨x, m₁, m₂, y, h⟩ := by simpa using comp_closure.intro x m₁ m₂ (y ≫ g) h
+
+/-- Hom-sets of the quotient category. -/
+def hom (s t : quotient r) := quot $ @comp_closure C _ r s.to_C t.to_C
+
+instance (a : quotient r) : inhabited (hom r a a) := ⟨quot.mk _ (𝟙 a.to_C)⟩
+
+/-- Composition in the quotient category. -/
+def comp {a b c : quotient r} : hom r a b → hom r b c → hom r a c :=
```

Comment:
Here I think you should also use the double braces `{{a b c ...}}` so that you can just use `comp r` downstairs, instead of `@comp _ _ r`.